libflux/src/semantic/walk.rs

```diff
@@ -0,0 +1,997 @@
+            Expression::Object(ref mut e) => Node::ObjectExpr(e),
+            Expression::Member(ref mut e) => Node::MemberExpr(e),
+            Expression::Index(ref mut e) => Node::IndexExpr(e),
+            Expression::Binary(ref mut e) => Node::BinaryExpr(e),
+            Expression::Unary(ref mut e) => Node::UnaryExpr(e),
+            Expression::Call(ref mut e) => Node::CallExpr(e),
+            Expression::Conditional(ref mut e) => Node::ConditionalExpr(e),
+            Expression::StringExpr(ref mut e) => Node::StringExpr(e),
+            Expression::Integer(ref mut e) => Node::IntegerLit(e),
+            Expression::Float(ref mut e) => Node::FloatLit(e),
+            Expression::StringLit(ref mut e) => Node::StringLit(e),
+            Expression::Duration(ref mut e) => Node::DurationLit(e),
+            Expression::Uint(ref mut e) => Node::UintLit(e),
+            Expression::Boolean(ref mut e) => Node::BooleanLit(e),
+            Expression::DateTime(ref mut e) => Node::DateTimeLit(e),
+            Expression::Regexp(ref mut e) => Node::RegexpLit(e),
+        }
+    }
+    fn from_stmt(stmt: &'a mut Statement) -> Node {
+        match *stmt {
+            Statement::Expr(ref mut s) => Node::ExprStmt(s),
+            Statement::Variable(ref mut s) => Node::VariableAssgn(s),
+            Statement::Option(ref mut s) => Node::OptionStmt(s),
+            Statement::Return(ref mut s) => Node::ReturnStmt(s),
+            Statement::Test(ref mut s) => Node::TestStmt(s),
+            Statement::Builtin(ref mut s) => Node::BuiltinStmt(s),
+        }
+    }
+    fn from_string_expr_part(sp: &'a mut StringExprPart) -> Node {
+        match *sp {
+            StringExprPart::Text(ref mut t) => Node::TextPart(t),
+            StringExprPart::Interpolated(ref mut e) => Node::InterpolatedPart(e),
+        }
+    }
+    fn from_assignment(a: &'a mut Assignment) -> Node {
+        match *a {
+            Assignment::Variable(ref mut v) => Node::VariableAssgn(v),
+            Assignment::Member(ref mut m) => Node::MemberAssgn(m),
+        }
+    }
+}
+
+// Visitor defines a visitor pattern for walking the AST
+//
+// When used with the walk function, Visit will be called for every node
+// in depth-first order. After all children for a Node have been visted,
+// Done is called on that Node to signal that we are done with that Node.
+//
+// If Visit returns None, walk will not recurse on the children.
+pub trait Visitor: Sized {
+    // Visit is called for a node.
+    // The returned visitor will be used to walk children of the node.
+    // If visit returns None, walk will not recurse on the children.
+    fn visit(&self, node: &mut Node) -> Option<Self>;
+    // Nest gets called when the walk encounters a Block.
+    // By default it returns None.
+    // If some visitor is returned, that will be used to continue walking
+    // down the current Block.
+    fn nest(&self) -> Option<Self> {
+        None
+    }
```

Comment:
Nesting I believe is only ever used once in the Go code and it's used as a way to get around the limitations imposed by the default traversal algorithm. I would honestly like to see it removed unless it is deemed absolutely necessary.

In the Go code nesting is used by the type checker during constraint generation. It's purpose it to add a new stack frame to the type environment when visiting a function block. However a new stack frame must be added ***before*** we visit the block. As a result, type inference cannot use the visitor pattern to generate constraints.

Reply:
Got it